pkg/provisioner/ironic/dependencies.go

```diff
@@ -0,0 +1,112 @@
+package ironic
+
+import (
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/drivers"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+)
+
+const (
+	checkRequeueDelay = time.Second * 10
+)
+
+type ironicDependenciesChecker struct {
+	client    *gophercloud.ServiceClient
+	inspector *gophercloud.ServiceClient
+}
+
+func newIronicDependenciesChecker(client *gophercloud.ServiceClient, inspector *gophercloud.ServiceClient) *ironicDependenciesChecker {
+	return &ironicDependenciesChecker{
+		client:    client,
+		inspector: inspector,
+	}
+}
+
+func (i *ironicDependenciesChecker) IsReady() (result provisioner.Result, err error) {
+
+	ready, err := i.checkIronic()
+	if ready && err == nil {
+		ready, err = i.checkIronicInspector()
+	}
+
+	if !ready && err == nil {
+		result.Dirty = true
+		result.RequeueAfter = checkRequeueDelay
+	}
+
+	return result, err
+}
+
+func (i *ironicDependenciesChecker) hasUnexpectedResponseCode(err error) bool {
+
+	errType := reflect.TypeOf(err)
+
+	if errType.Kind() == reflect.Struct {
+		targetType := reflect.TypeOf(gophercloud.ErrUnexpectedResponseCode{})
+		for i := 0; i < errType.NumField(); i++ {
+			f := errType.Field(i)
+			if f.Anonymous && f.Type == targetType {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (i *ironicDependenciesChecker) checkEndpoint(client *gophercloud.ServiceClient) (ready bool, err error) {
+
+	// NOTE: Some versions of Ironic inspector returns 404 for /v1/ but 200 for /v1,
+	// which seems to be the default behavior for Flask. Remove the trailing slash
+	// from the client endpoint.
+	endpoint := strings.TrimSuffix(client.Endpoint, "/")
+
+	_, err = client.Get(endpoint, nil, nil)
+	ready = err == nil
+	if !ready && i.hasUnexpectedResponseCode(err) {
+		err = nil
+	}
+
+	return ready, err
+}
+
+func (i *ironicDependenciesChecker) checkIronic() (ready bool, err error) {
+	ready, err = i.checkEndpoint(i.client)
+	if ready && err == nil {
+		ready, err = i.checkIronicConductor()
+	}
+	return ready, err
+}
+
+func (i *ironicDependenciesChecker) checkIronicConductor() (ready bool, err error) {
+
+	pager := drivers.ListDrivers(i.client, drivers.ListDriversOpts{
+		Detail: false,
+	})
+	err = pager.Err
+
+	if err == nil {
```

Comment:
This check feels backwards. Shouldn't it return early with an error if err != nil?

Reply:
I think it's just a matter of style, personally I also tend to prefer avoiding early return pattern when possible to increase code readability - especially for bigger functions.

Reply:
Ok!